dsl_parser/rel_graph.py

```diff
@@ -778,16 +782,19 @@ def _build_scaling_groups_map(ctx, node_instance_ids, group):
     return scaling_groups_map
 
 
-def _node_instance_id(node_id, ctx):
-    new_node_instance_id = '{0}_{1}'.format(node_id, _generate_id())
-    while new_node_instance_id in ctx.node_instance_ids:
-        new_node_instance_id = '{0}_{1}'.format(node_id, _generate_id())
-    ctx.node_instance_ids.add(new_node_instance_id)
-    return new_node_instance_id
-
-
-def _generate_id(id_len=6):
-    return ''.join(choice(digits + ascii_lowercase) for _ in range(id_len))
+def _generate_node_instance_ids(node_id, ctx, new_instance_num, id_len=8):
+    existing_ids = ctx.existing_node_instance_ids.get(node_id, set())
+    max_ids = min(16**id_len, 2**40)
+    try:
+        # generate enough ids for potential existing ones to be extracted from
+        ids = sample(range(max_ids), len(existing_ids) + new_instance_num)
+    except ValueError:
+        raise RuntimeError(
+            "Failed generating node instance ids for {0} - total node "
+            "instance number exceeds {1}".format(node_id, max_ids))
+    ids = set(ids) - existing_ids
+    ids = ['{0}_%0{1}x'.format(node_id, id_len) % id for id in ids]
```

Comment:
```python
>>> '{0}_{1:0{2}}'.format('a', 1, 16)
'a_0000000000000001'
```
str.format actually does multi-step formatting like that (with kwargs if you prefer)